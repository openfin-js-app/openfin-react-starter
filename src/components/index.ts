// BEGIN_OF_AL_CLI_COMPONENTS
export {default as ClientCounter} from './openfin-starter/client/ClientCounter';
export {default as Header} from './openfin-starter/Header/Header';
export {default as HeaderLinks} from './openfin-starter/Header/HeaderLinks';
export {default as HeaderThemeSwitcher } from './openfin-starter/Header/HeaderThemeSwitcher'
export {default as Sidebar} from './openfin-starter/Sidebar/Sidebar';
export {default as SnackbarContent} from './openfin-starter/Snackbar/MySnackbarContent';
export {default as OfflineOverlay} from './openfin-starter/Overlays/OfflineOverlay';
// END_OF_AL_CLI_COMPONENTS
