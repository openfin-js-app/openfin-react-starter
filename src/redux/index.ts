import {combineReducers} from 'redux';
import {IApplicationState} from './application/types';
import {IConfigState} from './config/types';

import application from './application/reducer';
import config from './config/reducer';

export * from './application/actions';
export * from './application/types';
export * from './config/actions';
export * from './config/types';

export interface IRootState {
    application:IApplicationState,
    config:IConfigState,
}

export default combineReducers({
    application,
    config,
});