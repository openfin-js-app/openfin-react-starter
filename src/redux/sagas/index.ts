import { all, takeEvery, select, fork } from 'redux-saga/effects';

import applicationSaga from './application';

export default function* rootSaga(){

    let sagas = [applicationSaga()];

    yield all(sagas);

}