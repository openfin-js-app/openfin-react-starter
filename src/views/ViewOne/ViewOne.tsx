import * as React from 'react';

export default class ViewOne extends React.Component<any,any>{
    render(){
        return(
            <React.Fragment>
                <span>Viewone works</span>
            </React.Fragment>
        );
    }
}