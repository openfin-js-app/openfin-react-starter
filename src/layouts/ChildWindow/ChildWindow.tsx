import * as React from 'react';
import { useEffect } from 'react'
import { connect } from 'react-redux';
import { Switch, Route, Redirect } from 'react-router-dom';
import cx from 'classnames';
import Snackbar from '@material-ui/core/Snackbar';

import { WithStyles ,withStyles} from '@material-ui/core/styles';

import {
    // acitons
    applicationToogleWindowState,
    applicationSetSnackbarStatus, applicationProcessSnackbarQueue,
    applicationCloseSnackbar,
    // types
    IRootState, ISnackBarMsg, MuiTheme,
} from '../../reduxs';


import childWindowRoutes from '../../routes/ChildWindow';

import { Header,SnackbarContent } from '../../components';

import { dashboardLayoutStyle as style } from '../../assets/jss/openfin-starter';
import {Window} from "redux-openfin";

declare const window:any;

const switchRoutes = (
    <Switch>
        {childWindowRoutes.map((prop:any,key)=>{
            if (prop.redirect)
                return <Redirect from={prop.path} to={prop.to} key={key}/>;
            return <Route path={prop.path} component={prop.component} key={key}/>

        })}
    </Switch>
);

interface IProps extends WithStyles<typeof style>{
    docked:boolean,
    snackBarOpen:boolean,
    snackBarMsgInfo:Partial<ISnackBarMsg>,
    windowsState:string,
    theme:MuiTheme,
    actions:{
        handleSetAsForeground: ()=> void,
        handleSnackbarClose: (event:any,reason:string)=> void,
        handleSnackbarCloseBtnClick: ()=> void,
        handleSnackbarExited: ()=> void,
        handleUndock: ()=> void,
        handleMinimize: ()=> void,
        handleMaximize: ()=> void,
        handleClose: ()=> void,
    }
}

const ChildWindowLayout:React.FunctionComponent<IProps> = (
    {
        classes,
        docked,snackBarOpen, snackBarMsgInfo, windowsState, theme,
        actions:{
            handleSetAsForeground,
            handleSnackbarClose, handleSnackbarCloseBtnClick, handleSnackbarExited,
            handleUndock, handleMinimize, handleMaximize, handleClose,
        },
        ...rest
    }
) => {

    useEffect(()=>{
        handleSetAsForeground();
        // return ()=>{
        //
        // }
    });

    return(
        <React.Fragment>
            <Header
                routes={childWindowRoutes}
                windowsState={windowsState}
                color={'info'}
                docked={docked}
                onUndock = {handleUndock}
                onMinimize={handleMinimize}
                onMaximize={handleMaximize}
                onClose = {handleClose}
                {...rest}
            />
            <div className={cx(classes.wrapper, classes.wrapperInfo)}>
                <div className={cx(
                    classes.mainPanel,classes.mainPanelShift
                )}
                >
                    <div className={classes.container}>
                        <div className={
                            cx(
                                classes.content,
                                {
                                    [classes.lightBoxShaddow]: theme === MuiTheme.LIGHT
                                }
                            )
                        }>
                            {switchRoutes}
                        </div>
                    </div>
                </div>
            </div>
            <Snackbar
                key={snackBarMsgInfo.key}
                anchorOrigin={{
                    vertical:'bottom',
                    horizontal:'center'
                }}
                open={snackBarOpen}
                autoHideDuration={6000}
                onClose={handleSnackbarClose}
                onExited={handleSnackbarExited}
            >
                <SnackbarContent
                    onClose={handleSnackbarCloseBtnClick}
                    variant={snackBarMsgInfo.variant}
                    message={snackBarMsgInfo.message}
                />
            </Snackbar>
        </React.Fragment>
    );
}

export default connect(
    (state:IRootState) => ({
        docked:state.application.docked,
        snackBarOpen:state.application.snackBarOpen,
        snackBarMsgInfo:state.application.snackBarMsgInfo,
        windowsState:state.application.windowsState,
        theme:state.config.application.theme,
    }),
    dispatch => ({
        actions:{
            handleSetAsForeground: () => {dispatch(Window.actions.setAsForeground({}))},
            handleSnackbarClose: (event,reason) => {dispatch(applicationCloseSnackbar({event,reason}))},
            handleSnackbarCloseBtnClick: () => {dispatch(applicationSetSnackbarStatus({open:false}))},
            handleSnackbarExited: () => {dispatch(applicationProcessSnackbarQueue())},
            // openfin
            handleUndock : () => {dispatch(Window.actions.leaveGroup({}))},
            handleMinimize: ()=>{dispatch(Window.actions.minimize({}))},
            handleMaximize: ()=>{dispatch(applicationToogleWindowState())},
            handleClose:()=>{dispatch(Window.actions.close({force:false}))},
        }
    })
)(withStyles(style)(ChildWindowLayout));