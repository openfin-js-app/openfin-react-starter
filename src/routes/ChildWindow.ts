import AccessibilityIcon from '@material-ui/icons/Accessibility';
import BusinessIcon from '@material-ui/icons/Business';
import ChatIcon from '@material-ui/icons/Chat';
import ReportIcon from '@material-ui/icons/Report';

import { RouteItem } from './Base';

import Accessibility from '../views/Accessibility/Accessibility';
import ViewOne from '../views/ViewOne/ViewOne';
import ViewTwo from '../views/ViewTwo/ViewTwo';
import ReportView from '../views/ReportView/ReportView';

const childRoutes:RouteItem[]=[
    {
        path:'/childWindow/accessibility',
        sidebarName:"Accessibility",
        navbarName:"View accessibility",
        icon: AccessibilityIcon,
        component: Accessibility,
    },
    {
        path:'/childWindow/view-one',
        sidebarName:"ViewOne",
        navbarName:"View one",
        icon: BusinessIcon,
        component: ViewOne,
    },
    {
        path:'/childWindow/view-two',
        sidebarName:"ViewTwo",
        navbarName:"View two",
        icon: ChatIcon,
        component: ViewTwo,
    },
    {
        path:'/childWindow/report',
        sidebarName:"Report",
        navbarName:"Desktop runtime report",
        icon: ReportIcon,
        component: ReportView,
    },
];

export default childRoutes;