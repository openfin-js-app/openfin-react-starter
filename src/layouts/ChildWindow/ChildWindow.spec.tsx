import * as React from 'react';
import { Provider } from 'react-redux';
import { createShallow, createMount } from '@material-ui/core/test-utils';
import configurestore from 'redux-mock-store';
import { MemoryRouter } from 'react-router';

import ChildWindow from './ChildWindow';
import ViewOne from '../../views/ViewOne/ViewOne';
import { Header } from '../../components';
import { rootDefaultState } from '../../redux'

const mockStore = configurestore();


const store = mockStore(rootDefaultState);

declare const window:any;

describe('ChildWindow layout',()=>{

    let shallow;
    let mount;

    beforeAll(() => {
        mount = createMount();
    });

    afterAll(() => {
        mount.cleanUp();
    });


    beforeEach(() => {
        shallow = createShallow();
    });

    it('render correctly on normal window state',()=>{
        let lastWindowState = 'normal';
        const currentWindow = {
            setAsForeground:jest.fn(),
            getState:jest.fn((cb)=>(cb(lastWindowState))),
            minimize:jest.fn(()=>{
                lastWindowState = 'minimized'
            }),
            restore:jest.fn(()=>{
                lastWindowState = 'normal'
            }),
            maximize:jest.fn(()=>{
                lastWindowState = 'maximized'
            }),
        };

        window.fin={
            desktop:{
                Window:{
                    getCurrent: ()=> currentWindow
                }
            }
        };

        const wrapper = mount(
            <Provider store={store}>
                <MemoryRouter initialEntries={['/childWindow/view-one']}>
                    <ChildWindow store={store} location={{pathname:'/childWindow/view-one'}}/>
                </MemoryRouter>
            </Provider>
        );
        expect(wrapper.find(ViewOne)).toHaveLength(1);
        expect(wrapper.find(ChildWindow)).toHaveLength(1);
        expect(wrapper.find(Header)).toHaveLength(1);
        wrapper.find(Header).props().onMaximize();
        expect(currentWindow.maximize).toHaveBeenCalled();
        expect(currentWindow.getState).toHaveBeenCalled();
    });

    it('could be minimized',()=>{
        let lastWindowState = 'normal';
        const currentWindow = {
            setAsForeground:jest.fn(),
            getState:jest.fn((cb)=>(cb(lastWindowState))),
            minimize:jest.fn(()=>{
                lastWindowState = 'minimized'
            }),
            restore:jest.fn(()=>{
                lastWindowState = 'normal'
            }),
            maximize:jest.fn(()=>{
                lastWindowState = 'maximized'
            }),
        };

        window.fin={
            desktop:{
                Window:{
                    getCurrent: ()=> currentWindow
                }
            }
        };

        const wrapper = mount(
            <Provider store={store}>
                <MemoryRouter initialEntries={['/childWindow/view-one']}>
                    <ChildWindow store={store} location={{pathname:'/childWindow/view-one'}}/>
                </MemoryRouter>
            </Provider>
        );
        expect(wrapper.find(ViewOne)).toHaveLength(1);
        expect(wrapper.find(ChildWindow)).toHaveLength(1);
        expect(wrapper.find(Header)).toHaveLength(1);
        wrapper.find(Header).props().onMinimize();
        expect(currentWindow.minimize).toHaveBeenCalled();
        expect(currentWindow.getState).toHaveBeenCalled();
    });

    it('could be closed',()=>{
        let lastWindowState = 'normal';
        const currentWindow = {
            setAsForeground:jest.fn(),
            getState:jest.fn((cb)=>(cb(lastWindowState))),
            minimize:jest.fn(()=>{
                lastWindowState = 'minimized'
            }),
            restore:jest.fn(()=>{
                lastWindowState = 'normal'
            }),
            maximize:jest.fn(()=>{
                lastWindowState = 'maximized'
            }),
            close:jest.fn(),
        };

        window.fin={
            desktop:{
                Window:{
                    getCurrent: ()=> currentWindow
                }
            }
        };

        const wrapper = mount(
            <Provider store={store}>
                <MemoryRouter initialEntries={['/childWindow/view-one']}>
                    <ChildWindow store={store} location={{pathname:'/childWindow/view-one'}}/>
                </MemoryRouter>
            </Provider>
        );
        expect(wrapper.find(ViewOne)).toHaveLength(1);
        expect(wrapper.find(ChildWindow)).toHaveLength(1);
        expect(wrapper.find(Header)).toHaveLength(1);
        wrapper.find(Header).props().onClose();
        expect(currentWindow.close).toHaveBeenCalled();
        expect(currentWindow.getState).toHaveBeenCalled();
    });

    it('render correctly on maximized window state',()=>{
        let lastWindowState = 'maximized';
        const currentWindow = {
            setAsForeground:jest.fn(),
            getState:jest.fn((cb)=>(cb(lastWindowState))),
            minimize:jest.fn(()=>{
                lastWindowState = 'minimized'
            }),
            restore:jest.fn(()=>{
                lastWindowState = 'normal'
            }),
            maximize:jest.fn(()=>{
                lastWindowState = 'maximized'
            }),
        };

        window.fin={
            desktop:{
                Window:{
                    getCurrent: ()=> currentWindow
                }
            }
        };

        const wrapper = mount(
            <Provider store={store}>
                <MemoryRouter initialEntries={['/childWindow/view-one']}>
                    <ChildWindow store={store} location={{pathname:'/childWindow/view-one'}}/>
                </MemoryRouter>
            </Provider>
        );
        expect(wrapper.find(ViewOne)).toHaveLength(1);
        expect(wrapper.find(ChildWindow)).toHaveLength(1);
        expect(wrapper.find(Header)).toHaveLength(1);
        wrapper.find(Header).props().onMaximize();
        expect(currentWindow.restore).toHaveBeenCalled();
        expect(currentWindow.getState).toHaveBeenCalled();
    });

});