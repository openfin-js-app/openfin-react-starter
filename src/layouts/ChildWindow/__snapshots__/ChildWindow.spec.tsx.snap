// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ChildWindow layout render SnackbarContent and could be closed correctly 1`] = `
Array [
  Object {
    "error": false,
    "payload": Object {},
    "type": "REDUX_OPENFIN::SET_AS_FOREGROUND",
  },
  Object {
    "error": false,
    "payload": Object {},
    "type": "REDUX_OPENFIN::MINIMIZE",
  },
  Object {
    "type": "APPLICATION_TOGGLE_WINDOW_STATE",
  },
  Object {
    "error": false,
    "payload": Object {
      "force": false,
    },
    "type": "REDUX_OPENFIN::CLOSE",
  },
  Object {
    "payload": Object {
      "open": false,
    },
    "type": "APPLICATION_SET_SNACKBAR_STATUS",
  },
]
`;

exports[`ChildWindow layout render in normal state correctly by default 1`] = `
Array [
  Object {
    "error": false,
    "payload": Object {},
    "type": "REDUX_OPENFIN::SET_AS_FOREGROUND",
  },
  Object {
    "error": false,
    "payload": Object {},
    "type": "REDUX_OPENFIN::MINIMIZE",
  },
  Object {
    "type": "APPLICATION_TOGGLE_WINDOW_STATE",
  },
  Object {
    "error": false,
    "payload": Object {
      "force": false,
    },
    "type": "REDUX_OPENFIN::CLOSE",
  },
  Object {
    "payload": Object {
      "event": undefined,
      "reason": undefined,
    },
    "type": "APPLICATION_CLOSE_SNACKBAR",
  },
  Object {
    "type": "APPLICATION_PROCESS_SNACKBAR_QUEUE",
  },
]
`;
